File: frontend/src/api/classImport.ts
import { api, AI_TIMEOUT } from './client';

export type PreviewCourse = {
  class_name: string;
  section: string | null;
  teacher_name: string;
  teacher_email: string | null;
  google_classroom_id: string | null;
  existing: boolean;
  existing_course_id: number | null;
};

export type PreviewConnected = {
  connected: true;
  courses: PreviewCourse[];
  error?: string;
};

export type PreviewDisconnected = {
  connected: false;
  connect_url: string;
  courses: [];
  error?: string;
};

export type PreviewResponse = PreviewConnected | PreviewDisconnected;

export type ParsedRow = {
  class_name: string;
  section: string | null;
  teacher_name: string;
  teacher_email: string | null;
};

export type ParseScreenshotResponse = {
  parsed: ParsedRow[];
};

export type BulkCreateRow = {
  class_name: string;
  section: string | null;
  teacher_name: string;
  teacher_email: string | null;
  google_classroom_id: string | null;
};

export type BulkCreatedItem = {
  index: number;
  course_id: number;
  name: string;
};

export type BulkFailedItem = {
  index: number;
  error: string;
  existing_course_id?: number;
};

export type BulkCreateResponse = {
  created: BulkCreatedItem[];
  failed: BulkFailedItem[];
};

export async function fetchGoogleClassroomPreview(): Promise<PreviewResponse> {
  const { data } = await api.get<PreviewResponse>('/api/courses/google-classroom/preview');
  return data;
}

export async function parseScreenshot(file: File): Promise<ParseScreenshotResponse> {
  const form = new FormData();
  form.append('image', file);
  const { data } = await api.post<ParseScreenshotResponse>(
    '/api/courses/parse-screenshot',
    form,
    {
      ...AI_TIMEOUT,
      headers: { 'Content-Type': 'multipart/form-data' },
    },
  );
  return data;
}

export async function bulkCreateClasses(rows: BulkCreateRow[]): Promise<BulkCreateResponse> {
  const { data } = await api.post<BulkCreateResponse>('/api/courses/bulk', { rows });
  return data;
}

File: app/schemas/class_import.py
"""Schemas for the CB-ONBOARD-001 class-import flow (#3985).

Bulk-create classes+teachers from Google Classroom or a parsed screenshot.
"""
from typing import Optional

from pydantic import BaseModel, EmailStr, Field, field_validator

from app.schemas.user import strip_whitespace


class BulkCreateRow(BaseModel):
    """One class/teacher row to create during bulk import."""

    class_name: str = Field(min_length=1, max_length=200)
    section: Optional[str] = Field(default=None, max_length=200)
    teacher_name: str = Field(min_length=1, max_length=200)
    teacher_email: Optional[EmailStr] = Field(default=None)
    google_classroom_id: Optional[str] = Field(default=None, max_length=255)

    @field_validator("class_name", "section", "teacher_name", "google_classroom_id", mode="before")
    @classmethod
    def _strip_whitespace(cls, v: object) -> object:
        return strip_whitespace(v)


class BulkCreateRequest(BaseModel):
    rows: list[BulkCreateRow] = Field(min_length=1, max_length=50)


class BulkCreateResult(BaseModel):
    """Top-level response from POST /bulk."""

    created: list[dict]
    failed: list[dict]


class ParsedScreenshotRow(BaseModel):
    class_name: str
    section: Optional[str] = None
    teacher_name: str
    teacher_email: Optional[str] = None


class ParseScreenshotResponse(BaseModel):
    parsed: list[ParsedScreenshotRow]


class GoogleClassroomPreviewCourse(BaseModel):
    class_name: str
    section: Optional[str] = None
    teacher_name: Optional[str] = None
    teacher_email: Optional[str] = None
    google_classroom_id: str
    existing: bool = False
    existing_course_id: Optional[int] = None


class GoogleClassroomPreviewResponse(BaseModel):
    connected: bool
    connect_url: Optional[str] = None
    error: Optional[str] = None
    courses: list[GoogleClassroomPreviewCourse] = Field(default_factory=list)
